### DISCUSSION TITLE
How to use new .NET 9 string.Join(char, ReadOnlySpan<string?>) with ReadOnlySpan<string>?

### DISCUSSION THREAD

**Discussion by kindermannhubert** [dotnet/runtime#109829] [Q&A] [answered]
Hi,

what is the expected way to use APIs like
```C#
string.Join(char, ReadOnlySpan<string?>)
```
 when I have `ReadOnlySpan<string>` in hand (which I expect would be much more often than `ReadOnlySpan<string?>`)?
 
 The problem I'm having is the following example:
```C#
void Test(ReadOnlySpan<string> span)
{
    var joined = string.Join(';', span);
    ...
}
```
generates warning
> Argument of type 'ReadOnlySpan\<string\>' cannot be used for parameter 'value' of type 'ReadOnlySpan\<string?\>' in 'string string.Join(char separator, params ReadOnlySpan\<string?\> value)' due to differences in the nullability of reference types.

I understand the problem but I'm wondering if there is some simple way to overcome this issue (I don't consider suppressing the warning here as a good solution). If not, I'm afraid the usability of such an API is severely harmed.

**Comment by huoyaoyuan** [comment, marked as answer]
The warning disappears in `<LangVersion>preview</LangVersion>`, the new first-class span feature should allow nullability covariance for spans.

Meanwhile, you can use `ReadOnlySpan<string?>.CastUp` to workaround this.

**Reply by kindermannhubert** [reply]
Great, thank you!
